### PR TITLE
Add ZED dependencies

### PR DESCRIPTION
**Pull request by wraftus**
# Pull Request

## Contribution Checklist

- [x] I have read the [contributing](https://github.com/uwreact/uwreact_robot/blob/master/CONTRIBUTING.md) guide.
- [x] I have referenced relevant issues.
- [x] I have labeled accordingly.
- [x] I have detailed my changes as much as possible.

## Change Checklist

- [ ] I have added tests.
- [x] I have added necessary documentation.
- [x] I have tested my changes on real hardware.

### CODE REVIEW

**Comment by wmmc88** [comment]
BUILD FAILED. Git gud

**Comment by wraftus** [comment]
oof, it needs CUDA and the ZED SDK to build and I'd imagine travis doesn't have that

**Comment by wmmc88** [comment]
Pretty sure it only needs CUDA and the sdk at runtime

**Comment by wraftus** [comment]
well catkin build fails unless you have them, unless that's something else

**Comment by wmmc88** [comment]
Oof ye nvm u need the sdk to build

**Comment by michaelwm** [comment]
I'd look into the travis docs and talk to @Matthew-Reynolds who set up the ROS CI to see if you can add the dependencies you need. If travis isn't enough, https://circleci.com will definitely be enough and we may have to look into transitioning sooner.

**Comment by michaelwm** [comment]
But most importantly, this *should not* be merged until CI works, no exceptions

**Comment by Matthew-Reynolds** [comment]
**tl;dr**: Blocked by frc_control PR #TODO and blocked by waiting for `ros-melodic-depthimage-to-laserscan` to be added to apt.

Travis vs Circle doesn't matter here at all. In both cases, we're running in a custom docker container. Since we're in normal docker, and not nvidia's special cuda-compatible `nvidia-docker`, trying to actually *use* CUDA will fail, but since we're only linking against CUDA and the ZED sdk, we're good to go.

ZED provides docker images for various configurations that already have CUDA and the ZED sdk installed. Doing some local testing, industrial-ci seems to have no issues building the zed ros code in their provided `zed:ubuntu1804-cuda10.0-zed2.7-ros` docker image. However, there are a few things blocking this PR.

1) The zed-ros-wrapper is written for Kinetic. Some of the examples require [depthimage_to_laserscan](https://github.com/ros-perception/depthimage_to_laserscan), which until 6 days ago, wasn't released in Melodic. It has since been added to the release index, so the next time the packages get updated, we should be in the clear for this. It is currently in the `ros-shadow-fixed` distribution, but not the main `ros` distribution. These get synced biweekly, but I don't know where in that 2 week cycle we are right now. But we are blocked until this package is available for install with apt. Workarounds for now are to either add `depthimage_to_laserscan` to our .rosinstall, or to only build with `ROS_REPO=ros-shadow-fixed` (We normally build with both `ros` and `ros-shadow-fixed`).

2) Using ubuntu18 for our CI breaks the build because `frc_control` hasn't merged its big update to 18 yet. However, frc_control's update is ready to be merged (PR #TODO), so this won't block us for long.

3) industrial-ci expects the user in the docker image to be root so that it can install packages and such. Zed's docker images, unlike the ones industrial-ci provides, don't have root as the default user. By manually editing the industrial-ci scripts, we can get around this, but this clearly isn't a viable workaround. I need to dig a bit deeper and figure out how to specify a user for the docker container through our .travis.yaml file.

**Comment by Matthew-Reynolds** [comment]
Side note, we should add a blocked tag

**Comment by Matthew-Reynolds** [comment]
Update on my prev comment: We are right at the beginning of the 2 week cycle.  Last sync was Feb 11, next will be ~Feb 25. I suggest our action for now is to remove our `ros` travis job, and only leave the `ros-shadow-fixed` job for the next 2 weeks, and then revert that change.

**Comment by Matthew-Reynolds** [comment]
Status update:

 - I implemented that workaround to only look at `ros-shadow-fixed` for now. #59 tracks the restoration of the standard `ros` repository. This should be done as soon as possible - High priority.
 - The issues I had with the docker images from StereoLabs have been resolved - in short, don't use their `-ros` images with industrial-ci.

This means the only thing blocking this PR is uwreact/frc_control#31. Once that has been merged, we can validate that things look good here, and then proceed with reviews.

File: README.md [hunk 1]
```diff
@@ -23,6 +23,29 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
...
+
+**ZED SDK**
+
+Go to: https://www.stereolabs.com/developers/release/#sdkdownloads_anchor and dowload the appropriate file
+To install run:
+```
+chmod +x ZED_SDK_{YOUR VERSION}.run
+./ZED_SDK_{YOUR VERSION}.run
+```
+and follow the command prompts
```
Comment (Matthew-Reynolds) [inline README.md]: (Maybe this is personal style, but) The line "and follow the command prompts" is kinda useless and just adds clutter. If they're running the script I think it's safe to assume they'll be following the prompts.

**Comment by Matthew-Reynolds** [comment]
uwreact/frc_control#31 is merged, rerunning CI.

**Comment by Matthew-Reynolds** [comment]
CI successful - PR is unblocked. @wmmc88 @michaelwm @wraftus Let's proceed with review.

File: README.md [hunk 2]
```diff
@@ -23,6 +23,29 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
 rosdep install --from-paths . --ignore-src -r -y
 ```
 
+In order to build anything zed related, you'll also need to install CUDA 10 and the ZED SDK
+
+**CUDA 10**
```
[the remark this answers is not in the document]
Comment (michaelwm) [reply README.md]: In this case, you should write 
### CUDA 10

```
### CUDA 10
```

Do the same number of `#` for below.

File: README.md [hunk 3]
```diff
@@ -23,6 +23,29 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
...
+
+Go to: https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=deblocal and follow install instruction according to your device & OS
+Once installed, add these to your ~/.bashrc file
```
Comment (michaelwm) [inline README.md]: ~/.bashrc should be in a code block, i.e. `~/.bashrc` using backticks `

File: README.md [hunk 4]
```diff
@@ -23,6 +23,29 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
 rosdep install --from-paths . --ignore-src -r -y
 ```
 
+In order to build anything zed related, you'll also need to install CUDA 10 and the ZED SDK
+
+**CUDA 10**
+
+Go to: https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=deblocal and follow install instruction according to your device & OS
```
[the remark this answers is not in the document]
Comment (michaelwm) [reply README.md]: Must make this a markdown link, saying something like:

[Download CUDA 10](https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=deblocal) and follow install instructions...

```
 [Download CUDA 10](https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=deblocal)
```

File: README.md [hunk 5]
```diff
@@ -23,6 +23,29 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
...
+Once installed, add these to your ~/.bashrc file
+```
+export PATH=$PATH:/usr/local/cuda-10.0/bin
+export CUDADIR=/usr/local/cuda-10.0
+export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda-10.0/lib64
+```
+
+**ZED SDK**
```
Comment (michaelwm) [inline README.md]: Same as above:

In this case, you should write 
### ZED SDK

```
### ZED SDK
```

File: README.md [hunk 6]
```diff
@@ -23,6 +23,29 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
...
+
+Go to: https://www.stereolabs.com/developers/release/#sdkdownloads_anchor and dowload the appropriate file
```
Comment (michaelwm) [inline README.md]: Same as above:

Must make this a markdown link, saying something like:

[Download the appropriate ZED SDK](https://www.stereolabs.com/developers/release/#sdkdownloads_anchor)

```
 [Download the appropriate ZED SDK](https://www.stereolabs.com/developers/release/#sdkdownloads_anchor)
```

File: README.md (hunk 1 above)
Comment (michaelwm) [reply README.md]: I agree with Matthew here

File: README.md [hunk 7]
```diff
@@ -23,6 +23,29 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
...
 ```
 
+In order to build anything zed related, you'll also need to install CUDA 10 and the ZED SDK
+
+**CUDA 10**
+
+Go to: https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=deblocal and follow install instruction according to your device & OS
+Once installed, add these to your ~/.bashrc file
+```
+export PATH=$PATH:/usr/local/cuda-10.0/bin
```
[the remark this answers is not in the document]
Comment (wraftus) [reply README.md]: I played around with it, and for our purposes I think your right. Just if you want to do any CUDA stuff directly, you need that first line to use the `nvcc` command

**Comment by wraftus** [comment]
Okay, finally got around to doing those documentation changes. Let me know if it's inline with what you guys want

File: README.md [hunk 8]
```diff
@@ -23,6 +23,20 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
 rosdep install --from-paths . --ignore-src -r -y
 ```
 
+In order to build anything zed related, you'll also need to install CUDA 10 and the ZED SDK
+
+### CUDA 10
+
+[Download CUDA 10](https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=deblocal) and follow install instructions according to which installation method you prefer
```
Comment (Matthew-Reynolds) [inline README.md]: I would still prefer to see this without the query string, but not a big deal. Approved either way.

Comment (wraftus) [reply README.md]: Sure, i can remove it, just thought I'd leave it for less confusion but its pretty straight forward

**Comment by Matthew-Reynolds** [comment]
@wmmc88 Please review so this PR can be merged.

File: README.md [hunk 9]
```diff
@@ -23,6 +23,21 @@ rosinstall --catkin . uwreact_robot/uwreact_robot.rosinstall
 rosdep install --from-paths . --ignore-src -r -y
 ```
 
+In order to build anything zed related, you'll also need to install CUDA 10 and the ZED SDK
+
+**CUDA 10**
```
[the remark this answers is not in the document]
Comment (Matthew-Reynolds) [reply README.md]: Don't recall, but since we're using Ubuntu 18 in our dev environment, the ZED SDK version requires CUDA 10.

**Comment by Matthew-Reynolds** [comment]
@wraftus I have changed this status to `urgent` since it is now blocking #64. Please acknowledge Melvin's comment and merge this PR ASAP.

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (wraftus) [reply README.md]: We're using CUDA 9 on the Jetson, but Matt is right, we need CUDA 10 to run the ubuntu18 SDK (which is different from the tx2 SDK even though we have 18 on the jetson)
